File: cpp_src/problem2-2-3.cpp
/**
 * @file      problem2-2-3.cpp
 * @author    zhangmin(3273749257@qq.com)
 * @brief     霍夫直线检测
 * @version   0.4
 * @date      2022-04-21
 * @copyright Copyright (c) 2022
 */



#include<iostream>
#include<opencv2/opencv.hpp>
#include<cmath>

using namespace std;
using namespace cv;



/**
* 函数名：     color2hsv
* 参数：       const Mat& color_img，需要转换为HSV图的原图像
* 返回值：     const Mat hsv_img，转换后的灰度图
* 函数功能：   将RGB彩色图转换为HSV图
*/
const Mat color2gray(const Mat& color_img)
{
    int size_h=color_img.rows;//获取像素长
    int size_w=color_img.cols;//获取像素宽

    //1.彩色图片通道分离
    vector<Mat> channels;
    split(color_img, channels);//B,G,R 11 59  30


    //2.构造一个单通道图
    Mat gray_img(size_h,size_w,CV_8UC1);

    //3.遍历每一个像素点
    for(int i=0;i<size_h;i++)
        for(int j=0;j<size_w;j++)
            //将rgb对应亮度值转化为灰度图对应亮度值
            gray_img.at<uchar>(i,j)=(uchar)((channels[0].at<uchar>(i,j)*11+channels[1].at<uchar>(i,j)*59+channels[2].at<uchar>(i,j)*30)/100);
    return gray_img;
}


//创建一个类进行高斯滤波
/**
* 类名：myGaussianFilter
* 类成员属性：Mat src，要进行高斯滤波的图像
*           int img_h，图像 像素长
*           int img_w，图像 像素宽
*           int gaussianSize_h，高斯核长
*           int gaussianSize_w，高斯核宽
*           double gaussianSigma，高斯核方差
* 类成员函数：
*           myGaussianFilter(Mat& _img,int gaussianSize_h,int gauaasianSize_w,double _sigma)，构造函数，初始化类成员属性
*           void getgaussianArray()，获取高斯核
*           Mat& gaussianMerge()，对图像进行高斯滤波
*/
class myGaussianFilter
{
private:
    /* 图像 */
    Mat src;//待处理的图像
    int img_h;//图像 像素长
    int img_w;//图像 像素长

    /* 高斯核 */
    Mat gaussianKernel;//高斯核
    int gaussianSize_h;//高斯核长
    int gaussianSize_w;//高斯核宽
    double gaussianSigma;//高斯核方差

public:
    //构造函数
    myGaussianFilter(Mat& _img,int h,int w,double _sigma):
    gaussianSigma(_sigma),gaussianSize_h(h),gaussianSize_w(w)
    {
        /* 图像 */
        src=_img;
        //获取图像长和宽
        img_h=src.rows;
        img_w=src.cols;

        /* 初始化高斯核 */
        //(1)获取高斯核的Size
        Size gaussianSize;
        gaussianSize.height=gaussianSize_h;
        gaussianSize.width=gaussianSize_w;
        //(2)创建高斯核
        gaussianKernel.create(gaussianSize,CV_64FC1);
        //(3)初始化高斯核(填充0)
        gaussianKernel=cv::Mat::zeros(gaussianSize,gaussianKernel.type());
    }
    Mat GaussianFilter();

private:
    void getgaussianKernel();
};
void myGaussianFilter::getgaussianKernel()
{
    // [1] 准备工作
    // [1-1] 获取中心坐标值
    int center_i= gaussianSize_h / 2;
    int center_j= gaussianSize_w / 2;
    // [1-2] 存储初步得出的高斯核的和
    double gaussianSum=0.0f;
    // [1-3] 存储坐标值
    double x, y;

    // [2] 初步计算高斯核
    // [2-1] 遍历每一个坐标
    for (int i = 0; i < gaussianSize_h; ++i)
    {
        y = pow(i - center_i, 2);
        for (int j = 0; j < gaussianSize_w; ++j)
        {
            // [2-2] 套公式计算每一个坐标的值
            x = pow(j - center_j, 2);
            //因为最后都要归一化的，常数部分可以不计算，也减少了运算量
            double g = exp(-(x + y) / (2 * gaussianSigma*gaussianSigma));
            gaussianKernel.at<double>(i, j) = g;
            gaussianSum += g;//求高斯核的和
        }
    }

    // [3] 归一化
    gaussianKernel = gaussianKernel / gaussianSum;
}
Mat myGaussianFilter::GaussianFilter()
{
    // [1] 准备工作
    // [1-1] 获取高斯核
    getgaussianKernel();
    // [1-2] 获取中心坐标值,为高斯滤波做准备
    int center_i = (gaussianKernel.rows - 1) / 2;
    int center_j = (gaussianKernel.cols - 1) / 2;
    // [1-3] 初始化一个图像，存放高斯滤波后的结果
    Mat dst = cv::Mat::zeros(src.size(),src.type());
    // [1-4] 边界填充src，存储到Newsrc中
    Mat Newsrc;
    copyMakeBorder(src, Newsrc, center_i, center_i, center_j, center_j, cv::BORDER_REPLICATE);//边界复制

    // [2] 高斯滤波
    // [2-1] 遍历每一个像素点
    for (int i = center_i; i < src.rows + center_i;++i)
    {
        for (int j = center_j; j < src.cols + center_j; ++j)
        {
           // [2-2] 卷积处理
           double sum[3] = { 0 };//存放卷积核与图像九宫格相乘的和(单通道或三通道)
           // 嵌套两层for循环遍历高斯掩膜
           for (int r = -center_i; r <= center_i; ++r)
           {
               for (int c = -center_j; c <= center_j; ++c)
               {
                   /* 通道数为1 */
                   if (src.channels() == 1)
                       sum[0] = sum[0] + Newsrc.at<uchar>(i + r, j + c) * gaussianKernel.at<double>(r + center_i, c + center_j);

                   /* 通道数为3 */
                   else if (src.channels() == 3)
                   {
                       // 利用Vec3b向量对三通道进行处理
                       Vec3b rgb = Newsrc.at<cv::Vec3b>(i+r,j + c);
                       sum[0] = sum[0] + rgb[0] * gaussianKernel.at<double>(r + center_i, c + center_j);//B
                       sum[1] = sum[1] + rgb[1] * gaussianKernel.at<double>(r + center_i, c + center_j);//G
                       sum[2] = sum[2] + rgb[2] * gaussianKernel.at<double>(r + center_i, c + center_j);//R
                   }
               }
           }

           // [2-3] 对值进行限制0-255
           for (int k = 0; k < src.channels(); ++k)
           {
               if (sum[k] < 0)
                   sum[k] = 0;
               else if (sum[k]>255)
                   sum[k] = 255;
           }


           // [2-4] 将处理完后的值放入对应通道
           /* 通道数为1 */
           if (src.channels() == 1)
               dst.at<uchar>(i - center_i, j - center_j) = static_cast<uchar>(sum[0]);

           /* 通道数为3 */
           else if (src.channels() == 3)
           {
               Vec3b rgb = { static_cast<uchar>(sum[0]), static_cast<uchar>(sum[1]), static_cast<uchar>(sum[2]) };//这里进行了类型转换
               dst.at<Vec3b>(i-center_i, j-center_j) = rgb;
           }
        }
    }

    // [3] 返回高斯模糊后的图像
    return dst;
}

/************************对图像进行边缘检测*************************/
/**
 * 函数名：  edgeDetection
 * 参数：   Mat src，需要进行边缘检测的图像
 * 返回值： Mat dst，边缘检测后的图像
 * 功能：   对图像进行边缘检测
 * 补充：   对单通道图像
 * ***************************************
 * 思路：
 * 1.主要准备工作
 *  (1)获取图像长、宽
 *  (2)创建一个图像，存储填充补零后的图像
 *  (3)创建一个图像，存储边缘处理后的图像
 *  (4)sobel算子，x方向、y方向
 *  (5)gx数组，gy数组，g
 *
 * 2.遍历每一个像素点,进行卷积处理（sobel算子）
 *  (1)
 *  (2)
 *  (3)
 *
 * 3.因为图像边缘不可能为边缘，故进行边缘置0
 *
 *
 * 3.返回边缘检测后的图像
 */

Mat edgeDetection(Mat src)
{
    // [1]
    // [1-1]
    int img_h=src.rows;
    int img_w=src.cols;
    // [1-2] 边界填充src，存储到Newsrc中
    Mat Newsrc;
    copyMakeBorder(src, Newsrc, 1, 1, 1, 1, cv::BORDER_REPLICATE);//边界复制一层
    // [1-3]
    Mat dst(img_h,img_w,CV_8UC1);
    // [1-4]
    int sobel_x[][3]={{-1,0,1},{-2,0,2},{-1,0,1}};
    int sobel_y[][3]={{-1,-2,-1},{0,0,0},{1,2,1}};
    // [1-5] gx、gy、g
    float gx[img_h][img_w],gy[img_h][img_w],g;

    // [2] 卷积处理(sobel算子)
    // [2-1] 遍历每一个像素点
    for(int i=1;i<img_h+1;i++)
        for(int j=1;j<img_w+1;j++)
        {
            for (int r = -1; r <= 1; ++r)
            {
                for (int c = -1; c <= 1; ++c)
                {
                     gx[i-1][j-1] += Newsrc.at<uchar>(i + r, j + c) * sobel_x[r+1][c+1];
                     gy[i-1][j-1] += Newsrc.at<uchar>(i + r, j + c) * sobel_y[r+1][c+1];
                }
                g=sqrt(gx[i-1][j-1]*gx[i-1][j-1]+gy[i-1][j-1]*gy[i-1][j-1]);

                if(g>255)
                    g=255;
                else if(g<0)
                    g=0;

                dst.at<uchar>(i-1,j-1)=g;
            }
        }


    // [3] 边缘置0
    for(int i=0;i<img_h;i++)
    {
        dst.at<uchar>(i,0)=0;
        dst.at<uchar>(i,img_w-1)=0;
    }
    for(int j=0;j<img_w;j++)
    {
        dst.at<uchar>(0,j)=0;
        dst.at<uchar>(img_h-1,j)=0;
    }

    // [4] 非极大值抑制
    // [4-1] 准备工作，变量定义
    uint8_t c,g1,g2,g3,g4;
    float weight,dTemp1,dTemp2;
    for(int i=1;i<img_h-1;i++)
    {
        for(int j=1;j<img_w-1;j++)
        {
            c=dst.at<uchar>(i,j);
            // [4-1] 判断gx、gy绝对值大小
            if(fabs(gx[i][j])<fabs(gy[i][j]))
            {
                weight= gx[i][j]/gy[i][j];
                g2=dst.at<uchar>(i-1,j);
                g4=dst.at<uchar>(i+1,j);

                //[4-2] 判断gx、gy是否同向
                if(gx[i][j]*gy[i][j]>0)
                {
                    g1=dst.at<uchar>(i-1,j-1);
                    g3=dst.at<uchar>(i+1,j+1);
                }

                else
                {
                    g1=dst.at<uchar>(i-1,j+1);
                    g3=dst.at<uchar>(i+1,j-1);
                }

                dTemp1 = weight * g1 + (1 - weight) * g2;
                dTemp2 = weight * g3 + (1 - weight) * g4;
            }

            else
            {
                weight= gy[i][j]/gx[i][j];
                g2=dst.at<uchar>(i,j-1);
                g4=dst.at<uchar>(i,j+1);

                //[4-2] 判断gx、gy是否同向
                if(gx[i][j]*gy[i][j]>0)
                {
                    g1=dst.at<uchar>(i+1,j-1);
                    g3=dst.at<uchar>(i-1,j+1);
                }

                else
                {
                    g1=dst.at<uchar>(i-1,j-1);
                    g3=dst.at<uchar>(i+1,j+1);
                }

                dTemp1 = weight * g1 + (1 - weight) * g2;
                dTemp2 = weight * g3 + (1 - weight) * g4;
            }

            if(c>=dTemp1&&c>=dTemp2);
            else
                dst.at<uchar>(i,j)=0;

        }
    }

    // [5] 边缘连接：使用双阈值
    int comp_max=200;
    int comp_min=50;
    // [5-1] 遍历像素点
    for(int i=1;i<img_h+1;i++)
        for(int j=1;j<img_w+1;j++)
        {
            // [5-2] 判断
            // [5-2-1] 大于comp_max
            if(dst.at<uchar>(i,j)>=comp_max);
            // [5-2-2] 小于comp_min
            else if(dst.at<uchar>(i,j))
                dst.at<uchar>(i,j)=0;
            // [5-2-3] 介于comp_min与comp_max之间
            else
            {

                // [5-3] 判断8邻域
                int flag=0;//设置一个标志位，来标志8邻域内是否存在边缘点，标志位为0不存在边缘点
                for(int k=-1;k<=1;k++)
                    for(int l=-1;l<=1;l++)
                        if(dst.at<uchar>(i+k,j+l)>=comp_max);
                            flag=1;
                 if(flag==0)
                 {
                     flag=0;//标志位清零
                     dst.at<uchar>(i,j)=0;
                 }
            }
        }


    return dst;
}



/**
* 函数名：     lineDetection
* 参数：       Mat src，绘画直线的图像
*             Mat edge_img，边缘检测后的图像
*             int voteThresh，是经过某一点曲线的数量的阈值，超过这个阈值，就表示这个交点所代表的参数对(rho, theta)在原图像中为一条直线
*             int stepRho = 1，距离步长，默认为1
*             int stepTheta =1，角度步长，默认为1
* 返回值：     无
* 函数功能：   对边缘图像进行直线检测
* ************************************************************
* 思路：
* 1.准备工作
*   (1)获取像素长，像素宽，图像对角线长
*   (2)霍夫变换后角度和距离的个数
*   (3)投票器
*
* 2.投票
*
* 3.阈值
*
* 4.非极大值抑制
*
* 5.输出直线
*/

void lineDetection(Mat src,Mat edge_img,int voteThresh,int stepRho = 1,int stepTheta =1)
{
    // [1] 准备工作
    // [1-1] 获取像素长，像素宽，图像对角线长
    int img_h=edge_img.rows;
    int img_w=edge_img.cols;
    int L=sqrt(img_h*img_h+img_w*img_w);
    // [1-2]
    int numRho=int(180.0 / stepTheta);
    int numTheta=int(2 * L / stepRho + 1);
    // [1-3] 投票器
    int accumulator[numRho][numTheta];
    // [1-4] 直线序列
    vector<vector<Point>> line_point(numRho*numTheta, vector<Point>(img_h*img_w));
 //   int line_total=0;//直线总数
    int pointnum=0;



    // [2] 投票
    // [2-1] 遍历除图像边界点以外的像素点
    for(int y=1;y<img_h-1;y++)
        for(int x=1;x<img_w-1;x++)
        {

            // [2-2] 遍历所有角点
            if(edge_img.at<uchar>(x,y)!=0)
            {
                for(int m=0;m<numTheta;m++)
                {
                    // [2-3] 对每一个角度，代入x、y，以及所有角点 计算对应的 rho 值
                    float rho = x *cos(stepTheta * m / 180.0 * CV_PI) + y * sin(stepTheta * m / 180.0 * CV_PI);
                    // [2-4] 计算投票哪一个区域
                    int n = int(round(rho + L) / stepRho);
                    // [2-5]投票加 1
                    accumulator[n][m]++;
                  int line_pos=m*(n+1);//m*n+m是对应位置，方便将line_pos解析W为一个m、n
                    line_point[line_pos][pointnum].x=x;//line_point[n][m],对应原空间的一条直线
                    line_point[line_pos][pointnum].y=y;

                }
                pointnum++;
            }
        }



    // [3] 阈值处理

    for(int i=1;i<numRho-1;i++)
        for(int j=1;j<numTheta-1;j++)
        {

            if(accumulator[i][j]>voteThresh&&accumulator[i][j]>=accumulator[i-1][j-1]&&accumulator[i][j]>=\
                    accumulator[i-1][j+1]&&accumulator[i][j]>=accumulator[i+1][j-1]&&accumulator[i][j]>=accumulator[i+1][j+1])
            {

                int pos=j*(i+1);


                line(src,line_point[pos][0],line_point[pos][pointnum-1],(0,0,255),2);

            }


        }



    imshow("line detection",src);
}



int main()
{
    // 1.读取图像
    Mat img=imread("D:/Picture/hough_line_detection.png");
    if(img.empty())//未获取到图像
    {
        cout << "can't read this image!" << endl;
        return 0;
    }

    // 2.实现高斯滤波
    myGaussianFilter test(img,3,3,sqrt(2));
    Mat gaus_img=test.GaussianFilter();
    // 3.灰度图
    Mat gray_img=color2gray(gaus_img);
    // 4.边缘检测
    Mat edge_img=edgeDetection(gray_img);

    lineDetection(img,edge_img,60);

    // 5.显示图像
    imshow("src",img);
    imshow("GaussianFilter",gaus_img);
    imshow("Gray",gray_img);
    imshow("Edge Detection",edge_img);

    // 6.霍夫直线检测
    lineDetection(img,edge_img,60);

    waitKey(0);


    return 0;
}

File: py_src/problem2-2-3.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2022/4/21 21:55
# @Author  : zhangmin
# @File    : prob2-2-3.py
# @Brief   : 霍夫直线检测
# @Version : 0.4

# -*- coding: utf-8 -*-
import numpy as np
import cv2
import math
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

'''
函数名：  color2gray
参数：    color_img，读取到的RGB彩色图
返回值：  gray_img，最终得出的灰度图
功能：    将RGB彩色图转换为灰度图
'''


def color2gray(color_img):
    size_h = color_img.shape[0]  # 获取图像宽像素
    size_w = color_img.shape[1]  # 获取图像长像素

    B = color_img[:, :, 0]  # 获取R通道--二维矩阵
    G = color_img[:, :, 1]  # 获取G通道--二维矩阵
    R = color_img[:, :, 2]  # 获取B通道--二维矩阵

    # 创建一个二维矩阵存储灰度图的亮度值
    gray_img = np.zeros((size_h, size_w), dtype=np.uint8)

    # 遍历所有像素点，套公式将R、G、B三通道的亮度值转换为灰度图的亮度值
    for i in range(size_h):
        for j in range(size_w):
            gray_img[i, j] = round((R[i, j] * 30 + G[i, j] * 59 + \
                                    B[i, j] * 11) / 100)

    return gray_img  # 返回灰度图


'''
函数名：    gaussian_filter
参数：     image          原图像
          K_size         高斯核尺寸，size*size
          sigma          高斯核方差
返回值：    gaussian_img   高斯滤波后的图像
功能：     对图像进行高斯滤波
'''


def gaussian_filter(img, K_size=3, sigma=2):
    # [1]获取图像长，宽，通道数
    H, W, C = img.shape

    # [2]图像填充，补零操作
    # [2-1]计算需要添加的层数，同时也方便后面的卷积操作
    center = K_size // 2
    # [2-2]创建一个三维矩阵，存放高斯滤波后的图像
    gaussian_img = np.zeros((H + center * 2, W + center * 2, C), dtype=np.float)
    # [2-3]将原图像copy到该矩阵中
    gaussian_img[center: center + H, center: center + W] = img.copy().astype(np.float)

    # [3]高斯核计算
    # [3-1]初始化高斯核，大小为size*size的二维矩阵
    gaussian_kernel = np.zeros((K_size, K_size), dtype=np.float)
    # [3-2]将坐标代入，初步计算出高斯核
    for x in range(-center, center):
        for y in range(-center, center):
            gaussian_kernel[y + center, x + center] = np.exp(-(x ** 2 + y ** 2) / (2 * (sigma ** 2)))
    gaussian_kernel /= (2 * np.pi * sigma * sigma)
    # [3-3]归一化
    gaussian_kernel /= gaussian_kernel.sum()

    # [4]进行高斯滤波
    # [4-1]将gaussian_img复制一份，方便卷积操作
    tmp = gaussian_img.copy()
    # [4-2]遍历所有像素点(三通道)，卷积操作
    for x in range(H):
        for y in range(W):
            for c in range(C):
                gaussian_img[center + x, center + y, c] = np.sum(gaussian_kernel * tmp[x: x + K_size, y: y + K_size, c])
    # [4-3]对矩阵的元素值进行范围限定，0~255
    gaussian_img = np.clip(gaussian_img, 0, 255)
    # [4-4]将矩阵元素类型转换为uint8
    gaussian_img = gaussian_img[center: center + H, center: center + W].astype(np.uint8)

    # [5]返回高斯滤波后的图像
    return gaussian_img


'''
函数名：    edgeDetect
参数：     src          原图像(单通道)   
          bmin=100     双阈值进行边界点判断 
          bmax=220     双阈值进行边界点判断
返回值：    dst         边缘检测后的图像(单通道)
功能：     对图像进行边缘检测
**************************************
思路：
    1.准备工作：
      (1)获取图像长，宽
      (2)将原图像copy一份并填充0
      (3)创建一个dst存储边缘检测后的图像
      (4)sobel_x,sobel_y sobel算子
      (5)gx，gy二维矩阵，存储梯度值

    2.遍历像素点，卷积处理
      (1)计算出x方向和y方向的梯度值
      (2)得到幅值
      (3)存入图像矩阵

    3.非极大值抑制
      (1)遍历像素点
      (2)与梯度方向进行比较

    4.边缘连接
      主要使用双阈值，对介于较大值与较小值的点查看其8邻域内是否有边
'''


def edgeDetect(src):
    # [1] 准备工作
    # [1-1] 获取图像长，宽
    img_h = src.shape[0]
    img_w = src.shape[1]
    # [1-2] 将原图像copy一份，并在四周填充一层0
    Newsrc = np.zeros((img_h + 2, img_w + 2), dtype=np.float)
    Newsrc[1:img_h + 1, 1:img_w + 1] = src.copy().astype(np.float)  # 利用切片src.copy().astype(np.float)
    # [1-3] 创建一个dst存储边缘检测后的图像
    dst = src.copy().astype(np.float)
    # [1-4] sobel_x,sobel_y
    sobel_x = [[1, 0, -1], [2, 0, -2], [1, 0, -1]]
    sobel_y = [[1, 2, 1], [0, 0, 0], [-1, -2, -1]]

    # [2] 遍历每一个像素点，进行卷积处理
    for i in range(1, img_h + 1):
        for j in range(1, img_w + 1):
            # [2-1] 算出gx、gy
            gx = np.sum(Newsrc[i - 1:i + 2, j - 1: j + 2] * sobel_x)  # gx
            gy = np.sum(Newsrc[i - 1:i + 2, j - 1: j + 2] * sobel_y)  # gy
            # [2-2] 算出g
            g = abs(gx) + abs(gy)
            # [2-3] 将求出的值存入dst
            dst[i - 1, j - 1] = g
    # [2-4] 注意对矩阵元素值的限定
    dst = np.clip(dst, 0, 255)
    # [2-5] 将矩阵元素类型转换为uint8
    dst = dst.astype(np.uint8)

    # [3] 返回
    return dst


# 函数功能：将src图像值小于comp_val的置为0
def gray2(src, comp_val):
    img_h = src.shape[0]
    img_w = src.shape[1]

    dst = np.zeros((img_h, img_w), dtype=np.uint8)

    for i in range(img_h):
        for j in range(img_w):
            if src[i, j] <= comp_val:
                dst[i, j] = 0
            else:
                dst[i, j] = src[i, j]

    return dst


# 霍夫极坐标变换：直线检测
'''
函数名：    hough_line_detection
参数：     src          原图像,绘制直线的图像   
          edge_img     边缘检测后的图像
          voteThresh   是经过某一点曲线的数量的阈值，超过这个阈值，就表示这个交点所代表的参数对(rho, theta)在原图像中为一条直线
          stepTheta =1 角度步长，默认为1
返回值：   无
功能：     对边缘图像进行直线检测
**************************************
思路：
    1.准备工作：      
    2.投票计数      
    3.阈值处理
    4.画线    
    5.展示图像
'''
def hough_line_detection(image,voteThresh = 60, stepTheta=1, stepRho=1):
    # [1] 准备工作
    # [1-1] 获取图像长，宽
    rows, cols = image.shape
    # [1-2] 图像中可能出现的最大垂线的长度
    L = round(math.sqrt(pow(rows - 1, 2.0) + pow(cols - 1, 2.0))) + 1
    # [1-3] 初始化投票器
    numtheta = int(180.0 / stepTheta)
    numRho = int(2 * L / stepRho + 1)
    accumulator = np.zeros((numRho, numtheta), np.int32)
    # []-4建立字典
    accuDict = {}
    for k1 in range(numRho):
        for k2 in range(numtheta):
            accuDict[(k1, k2)] = []

    # [2] 投票计数
    # [2-1] 遍历像素点
    for y in range(rows):
        for x in range(cols):
            if (image[y][x] ==255):  # 不对值为0的点做霍夫变换
                for m in range(numtheta):
                    # [2-2] 对每一个角度，计算对应的 rho 值
                    rho = x * math.cos(stepTheta * m / 180.0 * math.pi) + y * math.sin(stepTheta * m / 180.0 * math.pi)
                    # [2-3] 计算投票哪一个区域
                    n = int(round(rho + L) / stepRho)
                    # [2-4] 投票加 1
                    accumulator[n, m] += 1
                    # [2-5] 记录该点
                    accuDict[(n, m)].append((x, y))

    # 计数器大小
    rows, cols = accumulator.shape

    # [3] 阈值处理
    for r in range(rows):
        for c in range(cols):
            if accumulator[r][c] > voteThresh:
                points = accuDict[(r, c)]
                # [4] 画线
                cv2.line(img, points[0], points[len(points) - 1], (255), 2)


# 主函数
if __name__ == "__main__":
    # 1.读取图像
    img = cv2.imread("D:/Picture/hough_line_detection.png")

    # 2.高斯滤波
    gaussian_img = gaussian_filter(img, K_size=3, sigma=2)
    # 3.灰度图
    gray_img = color2gray(gaussian_img)
    # 4.边缘检测图
    edge_img = edgeDetect(gray_img)
    # 5.阈值后的图
    bw_img = gray2(edge_img, 30)

    # 6.霍夫直线检测
    hough_line_detection(bw_img, 1, 1)

    # 7.显示原图
    cv2.imshow("image", img)

    cv2.waitKey(0)
    cv2.destroyAllWindows()
